<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />



<meta name="progressive" content="false" />
<meta name="allow-skip" content="false" />
<meta name="learnr-version-prerender" content="0.11.4" />

<title>Multiple response variables</title>

<!-- header-includes START -->
<!-- HEAD_CONTENT -->
<!-- header-includes END -->
<!-- HEAD_CONTENT -->

<!-- highlightjs -->
<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>


<!-- taken from https://github.com/rstudio/rmarkdown/blob/de8a9c38618903627ca509f5401d50a0876079f7/inst/rmd/h/default.html#L293-L343 -->
<!-- tabsets -->
<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>
<!-- end tabsets -->


</head>

<body>
<a class='sr-only sr-only-focusable visually-hidden-focusable' href='#learnr-tutorial-content'>Skip to Tutorial Content</a>



<div class="pageContent band">
<main class="bandContent page">

<article class="topics" id="learnr-tutorial-content">

<div id="section-introduction" class="section level2">
<h2>Introduction</h2>
<div id="section-where-might-we-have-multiple-responses"
class="section level3">
<h3>Where might we have multiple responses?</h3>
<p>All the examples we have looked at so far have only had one response
variable, with one or more explanatory variables. However, there can be
numerous examples where you collect lots of response variables. Examples
include:</p>
<ul>
<li>invertebrate species x rocky shore quadrats</li>
<li>gene sequence x isolate number</li>
<li>metal concentrations x tissue sample</li>
<li>bacterial operational taxonomic units (OTUs) x sample</li>
<li>benthic species x trawl sample</li>
</ul>
<p>These data will normally be recorded as tables, where each column is
an <strong>attribute</strong>, and each row is a
<strong>sample</strong>. Therefore in the above examples, the attributes
are species, gene sequences and OTUs, whilst the samples are quadrats,
isolates, individual whose tissue was sampled etc. Often these tables of
data can be quite large, for example a typical invertebrate survey can
easily record over 50 species within one family, whilst genetic studies
can produce hundreds of sequences. Therefore the resulting tables can
produce large numbers of columns for the attributes, as well as the rows
for samples.</p>
<div class="panel panel-default tutorial-question-container">
<div data-label="why_multivariate" class="tutorial-question panel-body">
<div id="why_multivariate-answer_container" class="shiny-html-output"></div>
<div id="why_multivariate-message_container" class="shiny-html-output"></div>
<div id="why_multivariate-action_button_container" class="shiny-html-output"></div>
<script>if (Tutorial.triggerMathJax) Tutorial.triggerMathJax()</script>
</div>
</div>
</div>
<div id="section-summarising-your-data" class="section level3">
<h3>Summarising your data</h3>
<p>When you have lots of response variables, you want to have an easy
method to summarise your data, to help you interpret them. You might
want answers to questions such as:</p>
<ul>
<li>Which species of plants most frequently co-occur?</li>
<li>Which isolates are most similar in terms of their genetic
composition?</li>
<li>Which bacterial genes are associated with antibiotic
resistance?</li>
<li>Which quadrats have few species in common?</li>
</ul>
<p>In the past scientists used simple metrics to try and understand
these types of datasets. For example, species richness, which is the
number of species that occur in a sample. However, you can quickly see
that you could have two samples with the same number of species, but
few, or even no species in common. More sophisticated techniques
calculate <strong>diversity indices</strong> such as Shannon diversity
and Simpson diversity. These measure both the number of species (or
genes, or OTUs) and their frequency in the sample. However, again they
ignore the <strong>identity</strong> of these individual attributes:</p>
<ul>
<li>you might have two quadrats with the same Shannon diversity but few
species in common</li>
<li>you might have two isolates with similar Simpson diversity, but
little overlap in the gene sequences</li>
</ul>
</div>
<div id="section-displaying-similarities-between-attributes-and-samples"
class="section level3">
<h3>Displaying similarities between attributes and samples</h3>
<p>A better approach is to use methods that <strong>summarise the
composition</strong> of your samples. By composition, we are referring
to the number of species, their relative abundance, and the taxonomic
identity of those present in each sample. Likewise if your sample is an
isolate with gene frequencies, or OTUs etc. We can then measure the
<strong>similarity in their composition</strong> to make it easy to
compare samples or isolates.</p>
<p>This similarity in composition between samples can be measured on
a:</p>
<ul>
<li><strong>continuous scale</strong>, so that samples with similar
composition are placed close together, and very dissimilar samples
placed far apart.</li>
<li><strong>categorical scale</strong>, so samples with similar
composition are placed into different groups, classes or categories</li>
</ul>
<p>When we measure composition on a continuous scale it is referred to
as <strong>ordination</strong> and when we do it on a categorical scale
it is known as <strong>classification</strong> or
<strong>clustering</strong>. The two approaches can be considered
complementary.</p>
</div>
<div id="section-what-about-the-explanatory-variables"
class="section level3">
<h3>What about the explanatory variables?</h3>
<p>These can be used to help you interpret the results of your
ordination or classification, so that you have a proper biological
understanding of what is taking place. There are several ways of using
explanatory variables. The explanatory variables can be incorporated
into an ordination: this is known as <strong>constrained
ordination</strong> . Where the ordination is more descriptive, and does
not include explanatory variables, it is called <strong>unconstrained
ordination</strong>. (There are not constrained equivalents for
classification.) We will start with unconstrained ordination methods as
these are the simplest to introduce.</p>
</div>
</div>
<div id="section-unconstrained-ordination" class="section level2">
<h2>Unconstrained ordination</h2>
<p>The first biologists to struggle with multi-attribute data were
ecologists, because of the large numbers of species that they collected
in many field surveys. As a result, some of the methods will refer to
<strong>sites</strong> and <strong>species</strong>, but please remember
that the ‘sites’ could just as easily be samples, isolates, experimental
plants etc., whilst the ‘species’ could be OTUs, gene sequences etc.</p>
<div id="section-column-and-row-names" class="section level3">
<h3>Column and row names</h3>
<p>For all multi-attribute (also known as multivariate) datasets R
expects your raw data to have row names, as well as column names. These
can be alphnumeric, but column names must <strong>not</strong> start
with a number. If you save your data from Excel in .csv (comma separated
value) format, the <code>read.csv()</code> function will assume that the
first column contains row names. You will often want to keep row and
column names fairly short, so you can display them on the summary graphs
from your analyses. For example, typically species names are no longer
than 8- letter codes (4 letters genus, 4 letters species).</p>
</div>
<div id="section-dune-dataset" class="section level3">
<h3>Dune dataset</h3>
<p>We are going to analyse some vegetation data from the Netherlands,
which is considered a ‘classic’ dataset and is used in quite a wide
range of books.</p>
<p><img src="images/dunes.jpeg" /></p>
<p>This has been stored in a <code>data.frame</code> called
<code>dune</code>: use the exercise box below to get a summary of the
data. How many attributes (species) are present? How many samples
(quadrats)? What seems to be a common number in the data?</p>
<div class="tutorial-exercise" data-label="dune_summary"
data-completion="1" data-diagnostics="1" data-startover="1"
data-lines="0">
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<div class="tutorial-exercise-support"
data-label="dune_summary-solution" data-completion="1"
data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>summary(dune)
head(dune)
nrow(dune)
ncol(dune)
rownames(dune)
colnames(dune)</code></pre>
</div>
<p>You can see that the column names (attributes) are species, for
example “Achimill” is <em>Achillea millefolium</em> :</p>
<p><img src="images/achillea.jpeg" /> The rownames can be numbers, which
can be useful for graphical display. The dataset contains 20 quadrats
from <code>nrow()</code> and 30 species from <code>ncol()</code>. The
values shown in the <code>head()</code> function is the cover-abundance
(measured on a scale of 0 to 10), but notice how common the value 0 is
in these data. This is quite common, and known as a
<strong>sparse</strong> dataset.</p>
</div>
<div id="section-the-vegan-package" class="section level3">
<h3>The <code>vegan</code> package</h3>
<p>We are going to use the <code>vegan</code> package for these
analyses. It was originally developed for vegetation analysis (hence its
name), but has since been used in a wide range of other disciplines in
biology, including landscape analysis, molecular biology etc. If looking
for information on Google about the package, remember to add the word
‘multivariate’ to your search, otherwise it will return rather a lot of
recipes for food!</p>
</div>
<div id="section-different-types-of-unconstrained-ordination"
class="section level3">
<h3>Different types of unconstrained ordination</h3>
<p>There are several types of unconstrained ordination, and we will
restrict ourselves to three.</p>
<ul>
<li>principal components analysis (PCA) - a linear method</li>
<li>correspondence analysis (CA) - a unimodal method</li>
<li>non-metric multidimensional scaling (NMDS) - ranked similarity
method</li>
</ul>
<p>Linear methods such as PCA assume a relatively constant increase or
decrease in your attribute across your samples. So if some of your
samples had low soil pH, and some high pH, the abundance of an
individual species might increase or decline in roughly a straight line.
If the analysis works well, the PCA will create new variables that
‘capture’ most of this change:</p>
<p><img src="images/linear.png" /></p>
<p>Unimodal methods such as CA assume that the abundance of an attribute
can rise and fall along the axis. Where data are ‘sparse’ this can
sometimes be more realistic for the analysis:</p>
<p><img src="images/unimodal.png" /></p>
<p>NMDS uses a quite different approach. It ranks the abundance values
of your attributes (species, OTUs etc) and calculates pairwise
similarities. It then tries to simplify these sets of similarities
between all the original variables into 2 new variables for easy
visualisation. It can be useful when the original data are extremely
messy and difficult to interpret by PCA or CA.</p>
</div>
</div>
<div id="section-principal-components-analysis" class="section level2">
<h2>Principal components analysis</h2>
<p>Let’s analyse the dune data using principal components analysis
(PCA). The <code>vegan</code> package provides access to many advanced
multivariate methods. It has the <code>rda()</code> function, which when
used with only a response variable provides PCA, and if used with both
response and explanatory variables it automatically uses constrained
analysis. This can be a little confusing for beginners. Remember that in
both constrained and unconstrained analyses the “response variable” is
actually your <code>data.frame</code> containing lots of variables. The
<code>summary()</code> function on your ordination output produces lots
of information (most of which you do not need to study), so for this
website we’re going to limit amount of output. When you run it
interactively in RStudio, you’ll have to scroll up a bit to see the next
few lines.</p>
<pre class="r"><code># Undertake the PCA ordination
dune_pca &lt;- rda(dune)

# First part of summary: importance of derived variables
summary(dune_pca)</code></pre>
<pre><code>## Importance of components:
##                           PC1     PC2     PC3     PC4    PC5     PC6     PC7
## Eigenvalue            24.7953 18.1466 7.62913 7.15277 5.6950 4.33331 3.19936
## Proportion Explained   0.2947  0.2157 0.09069 0.08503 0.0677 0.05151 0.03803
## Cumulative Proportion  0.2947  0.5105 0.60115 0.68618 0.7539 0.80539 0.84342
##                           PC8    PC9    PC10    PC11    PC12    PC13     PC14
## Eigenvalue            2.78186 2.4820 1.85377 1.74712 1.31358 0.99051 0.637794
## Proportion Explained  0.03307 0.0295 0.02204 0.02077 0.01561 0.01177 0.007582
## Cumulative Proportion 0.87649 0.9060 0.92803 0.94880 0.96441 0.97619 0.983768
##                           PC15     PC16     PC17     PC18     PC19
## Eigenvalue            0.550827 0.350584 0.199556 0.148798 0.115753
## Proportion Explained  0.006548 0.004167 0.002372 0.001769 0.001376
## Cumulative Proportion 0.990316 0.994483 0.996855 0.998624 1.000000</code></pre>
<p>Creating the PCA is very simple. In a PCA you try to “squash” as much
of the variability in your data into new derived variables, known as
<strong>ordination axes</strong> where the first axis,
<strong>PC1</strong> “captures” as much of the variation as possible.
The second axis, <strong>PC2</strong> captures as much of the remaining
variation etc.</p>
<p>Therefore in the above output, you can see on the line labelled
<code>Proportion Explained</code> that PC1 captures 0.2947 = 29.5% of
the variation, whilst PC2 captures 0.2157 = 21.6% of the variation.
Together, shown in the line labelled <code>Cumulative Proportion</code>,
they jointly capture 0.5105 = 51.1% of the variation. In other words,
these two new derived axes summarise just over half the variation in
your original table of 30 attributes (species) by 20 samples (quadrats
or sites). Other axes are produced, but note that they rapidly become
less important, with PC3 explaining only 9% etc., so usually you only
look at PC1 and PC2.</p>
<div id="section-visualising-the-results" class="section level3">
<h3>Visualising the results</h3>
<p>PC1 and PC2 are created for both the attributes and samples, and the
easiest way to study them is via two graphs, that should be interpreted
together. You can display the results using the default
<code>plot()</code> command. When you have a lot of data these can
difficult to interpret.</p>
<p>The plot displays the species scores and site scores. Remember that
depending on your experimental data, the species scores might be what
you have measured so metal concentrations in a tissue sample or gene
sequences while the sites scores will correspond to the individual you
have sampled from.</p>
<pre class="r"><code># Display both on the same plot
plot(dune_pca)</code></pre>
<p><img src="Unconstrained_files/figure-html/plot_dune-1.png" width="624" /></p>
<p>The above plot is relatively straight forward to interpret. The
species scores have been plotted as text and site scores have been
plotted at numbers. There is a lot of information we can gather from
this plot:</p>
<ul>
<li>sampling locations with similar site scores will have similar
species compositions,</li>
<li>where species scores are close together, then they will often be
found at the same site,</li>
<li>the closer a species is to site, the higher its abundance,</li>
<li>the closer a species is to the centre of the plot (PC1 = 0 and PC2 =
0), the more common the species across all sites.</li>
</ul>
</div>
<div id="section-quick-interpretation-of-the-plots"
class="section level3">
<h3>Quick interpretation of the plots</h3>
<p>Even from these simple plots you can answer some questions. See if
you can work them out:</p>
<div class="panel panel-default tutorial-question-container">
<div data-label="simple_pca_interpret" class="tutorial-question panel-body">
<div id="simple_pca_interpret-answer_container" class="shiny-html-output"></div>
<div id="simple_pca_interpret-message_container" class="shiny-html-output"></div>
<div id="simple_pca_interpret-action_button_container" class="shiny-html-output"></div>
<script>if (Tutorial.triggerMathJax) Tutorial.triggerMathJax()</script>
</div>
</div>
<p>Hopefully you can see that species close together in the ordination
graphs tend to co-occur, and are found mainly in the quadrats in that
part of the graph. In PCA plots you will often get a mass of overlapping
species in the centre of the plot, along the zero-zero lines: these tend
to be fairly widespread and ubiquitous species, not characteristic of
any particular sites.</p>
<p>If we wanted to look at the those species in the centre of the plot,
we can look at the species scores more closely. We can extract the
species scores using the <code>scores()</code> function and select PC1
and PC2 with the <code>choices = 1:2</code>. Run the code below and look
down PC1 and PC2 columns to see if you can find Trifrepe
(<em>Trifolium</em> <em>repens</em>). What are the PC1 and PC2 scores?
Are there any other species which are found closer to PC1 = 0 and PC2 =
0?</p>
<div class="tutorial-exercise" data-label="sum_pca" data-completion="1"
data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>scores(dune_pca, display=&quot;species&quot;, choices = 1:2)</code></pre>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<p>Looking down the columns for PC1 and PC2, you should see that
Trifrepe (<em>Trifolium</em> <em>repens</em>) has a PC1 = 0.581 and PC2
= 0.021. This species is hard to decipher from the cloud of text in the
middle of the plot but given these values we would expect to find it
most of the sites. If we look at the data in <code>dune</code> we find
it at 16 out of the 20 sites. Sometimes it is worth looking directly at
the species scores to help interpret your plot. Are there any other
species which are found closer to PC1 = 0 and PC2 = 0?</p>
</div>
</div>
<div id="section-relate-pca-to-explanatory-variables"
class="section level2">
<h2>Relate PCA to explanatory variables</h2>
<div id="section-what-do-the-pca-axes-mean" class="section level3">
<h3>What do the PCA axes mean?</h3>
<p>For most people one of the confusing things about PCA and other
ordination techniques is that the PCA axes do not have any units, and
you are left wondering what they really ‘mean’. In essence, that is up
to you to find out, by trying to relate them to potential explanatory
data. Typically you only need to look at the first ordination axis (PC1)
as this captures most of the variation, but rarely you may want to look
at PC2 as well.</p>
</div>
<div id="section-environmental-data-for-the-sand-dune-vegetation"
class="section level3">
<h3>Environmental data for the sand dune vegetation</h3>
<p>We have some environmental data for the sand dune survey stored in a
<code>data.frame</code> called <code>dune.env</code>. Examine the data
in the box below:</p>
<div class="tutorial-exercise" data-label="check_env"
data-completion="1" data-diagnostics="1" data-startover="1"
data-lines="0">
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<div class="tutorial-exercise-support" data-label="check_env-solution"
data-completion="1" data-diagnostics="1" data-startover="1"
data-lines="0">
<pre class="text"><code>summary(dune.env)
head(dune.env)
nrow(dune.env)
ncol(dune.env)</code></pre>
</div>
<p>Notice that the <code>dune.env</code> dataset has the same number of
rows as <code>dune</code>, i.e. one entry for each quadrat sample. It
does not need to have the same number of columns. The particular
variables in the <code>dune.env</code> dataset are:</p>
<ul>
<li>A1 = thickness of soil A1 horizon</li>
<li>Moisture = ordered factor 1, 2, 4, 5</li>
<li>Management = factor where BF (biological farming), HF (hobby
farming), NM (nature conservation management), SF (standard farm
management)</li>
<li>Use = ordered factor with levels Hayfield &lt; Haypastu &lt;
Pasture</li>
<li>Manure = ordered factor 0 to 4</li>
</ul>
<p>Let’s look at the relationship between Moisture and the PC1 scores.
You can extract the scores from your <code>dune_pca</code> object using
the <code>scores()</code> function, so to get the site (quadrat or
sample) scores, use
<code>scores(dune_pca, display="sites", choice=1)</code>. The
<code>choice=1</code> indicates to just extract the PC1 scores. Use this
command, storing the results in <code>dune_pc1</code> and create a
boxplot against <code>Moisture</code> using
<code>gf_boxplot()</code>:</p>
<div class="tutorial-exercise" data-label="dune_moisture_boxplot"
data-completion="1" data-diagnostics="1" data-startover="1"
data-lines="0">
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<div class="tutorial-exercise-support"
data-label="dune_moisture_boxplot-solution" data-completion="1"
data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># Extract the PC1 site scores
dune_pc1 &lt;- scores(dune_pca, display=&quot;sites&quot;, choices = 1)

# Create boxplot
gf_boxplot(dune_pc1 ~ Moisture, data=dune.env)</code></pre>
</div>
<p>So, you can conclude that when PC1 is <strong>high</strong> the soil
moisture content is also relatively <strong>high</strong>. Now go back
to your plot of the species on PC1 and PC2.</p>
<div class="panel panel-default tutorial-question-container">
<div data-label="which_spp_like_moisture" class="tutorial-question panel-body">
<div id="which_spp_like_moisture-answer_container" class="shiny-html-output"></div>
<div id="which_spp_like_moisture-message_container" class="shiny-html-output"></div>
<div id="which_spp_like_moisture-action_button_container" class="shiny-html-output"></div>
<script>if (Tutorial.triggerMathJax) Tutorial.triggerMathJax()</script>
</div>
</div>
<p>In the same way that you have produced a graph for your Moisture,
investigate the other environmental variables against PC1. Remember,
depth of the soil A1 horizon is continuous, therefore
<code>gf_point()</code> is a more appropriate function.</p>
<div class="tutorial-exercise" data-label="check_other_variables"
data-completion="1" data-diagnostics="1" data-startover="1"
data-lines="0">
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<div class="tutorial-exercise-support"
data-label="check_other_variables-solution" data-completion="1"
data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>gf_point(dune_pc1 ~ A1, data=dune.env)
gf_boxplot(dune_pc1 ~ Management, data=dune.env)
gf_boxplot(dune_pc1 ~ Use, data=dune.env)
gf_boxplot(dune_pc1 ~ Manure, data=dune.env)</code></pre>
</div>
<p>Which variables seem to have little relationship with PC1, which ones
appear to go in different directions. If you have time, try and extract
PC2 scores and see how they relate to the environment.</p>
</div>
</div>
<div id="section-pca-arch-effect" class="section level2">
<h2>PCA arch effect</h2>
<div id="section-what-can-go-wrong-with-pca" class="section level3">
<h3>What can go wrong with PCA?</h3>
<p>Sometimes the characteristics of your original data can cause
problems for PCA, for example nearly all the samples squashed to one
side of the plot, with a couple at the other extreme. This can arise
when there are a couple of ‘outlier’ samples with an unusual composition
of attributes. Sometimes this can be corrected by standardising your
data before analysis:</p>
<p><code>mydata_std &lt;- decostand(mydata_raw, method="standardize")</code></p>
<p>which standardises all the attributes to have zero mean and unit
(1.0) standard deviation. Other standardisation methods are available in
the <code>decostand()</code> function, and the <code>"hellinger"</code>
option can also be quite useful. See <code>?decostand</code> for the
help pages.</p>
<p>Another problem is more subtle, and is known as the
<strong>arch</strong> or <strong>horseshoe</strong> effect. It seems to
be a particular problem for both ecological and gene sequencing data.
This is when the differences in the composition of the samples is so
extreme that PCA starts to pull the ends of the ordination together,
forming an arch. Often <strong>unimodal methods</strong> (see earlier)
such as CA or NMDS can correct this problem.</p>
</div>
<div id="section-arch-effect-example" class="section level3">
<h3>Arch effect example</h3>
<p>The following exercise uses data from Vare <em>et al</em> (1995) who
studied the effects of reindeer grazing in Scandinavia on the vegetation
in pine forests.</p>
<p><img src="images/reindeer.jpg" /></p>
<p>The data consist of cover estimates of 44 species of plants
(attributes, in columns) at 24 survey sites. You can read the study
<a href="www/JVG1995.pdf" target="_blank">here</a> If you look at the
paper you’ll notice that they used NMDS to analyse their data. What goes
wrong with PCA? Try it now, in the exercise box below; the data are
stored in the data.frame <code>varespec</code>:</p>
<div class="tutorial-exercise" data-label="archpca" data-completion="1"
data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># Create the PCA
varespec_pca &lt;- rda(varespec)

# Check variation explained; if using RStudio just enter summary(varespec_pca)
summary(varespec_pca)[[&quot;cont&quot;]][[&quot;importance&quot;]]

# Plot PC1 and PC2 site scores only
plot(varespec_pca, display = &quot;sites&quot;) 

#plot PC1 and PC2 species scores only
plot(varespec_pca, display = &quot;species&quot;)</code></pre>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<p>You can see that the amount of variation explained by PC1 and PC2 is
high, at 53.8% and 25.4%, so together they explain almost 80% of the
variation. This very high explained variation on just a couple of axes
is sometimes caused by an <strong>arch effect</strong>, which you can
clearly see in the resultant ordination plot. Notice too that in the
species plot, nearly all of them are squashed together at the centre,
with just 4 outlier species.</p>
<p>In this situation, we would need to think about other ordination
methods.</p>
</div>
</div>
<div id="section-correspondence-analysis-ca" class="section level2">
<h2>Correspondence analysis (CA)</h2>
<p>Correspondence analysis was the second of the three unconstrained
ordination methods we introduced at the start. CA is a weighted form of
PCA that can fit non-linear responses. We saw earlier that CA is used
when our attributes rise and fall along a gradient to create a unimodal
relationship. If our attributes are abundance or composition
measurements for each species, then the rise and fall of species along a
gradient represents <strong>beta-diversity</strong>. Beta-diversity
represent the change in species contribution between sites or along an
environmental gradient.</p>
<p>Let’s repeat the analysis using correspondence analysis (CA), which
assumes a <strong>unimodal</strong> relationship of the attributes (here
these are species) and the derived ordination axes. To undertake a
correspondence analysis, use the <code>cca()</code> function instead of
the <code>rda()</code> function.</p>
<div class="tutorial-exercise" data-label="ca1" data-completion="1"
data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># Create the CA
varespec_ca &lt;- cca(varespec)

# Check variation explained; if using RStudio just enter summary(varespec_pca)
summary(varespec_ca)

# plot CA1 and CA2 for samples and species
plot(varespec_ca)

# Plot CA1 and CA2 of the samples
plot(varespec_ca, display=&quot;sites&quot;) 

# Plot CA1 and CA2 of the species
plot(varespec_ca, display=&quot;species&quot;) </code></pre>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<p>The total amount of variation explained by the first two axes is much
less, at 25.2% and 17.1% respectively for CA1 and CA2, explaining 42.2%
in total (see the lines labelled <code>Proportion Explained</code> and
<code>Cumulative Proportion</code>). However, the graphs for both the
sites and species are much easier to understand, and the arch effect has
been removed.</p>
<p>With these data, CA would appear to be much better at analysing our
data but we mentioned beta-diversity earlier. What additional ecological
interpretation can we get from looking at the plot?</p>
<p>If we look at the <strong>site scores</strong> then we can see that
along CA1 (x-axis) there is a big difference between the community
composition at site 10 and 28. We would expect to see that the
contribution for each of the species examined would be different between
these two sites. The change in species composition from site to site
represents beta-diversity.</p>
<p>This can be illustrated more clearly by looking at the relationship
between the species cover estimates (original data collected) and CA1
scores. We can do this is four steps: extract site scores; select 4
species of interest; create a dataframe with sites scores and species
coverage estimates; and final plot those data.</p>
<div class="tutorial-exercise" data-label="ca3" data-completion="1"
data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># Extract just the CA1 scores
sco1 &lt;- scores(varespec_ca, display = &quot;sites&quot;, ch=1)

# look at the top 6 rows of data
head(sco1)

# Select four species of interest and plot against CA1
varspec2 &lt;- varespec  %&gt;%  select(Cladstel, Pleuschr, Vaccmyrt, Empenigr)

# Merge CA1 scores and species abundance into one table and turn the row names into a column so we can plot the site numbers on the figure
mydat &lt;- cbind(sco1, varspec2)
mydat$site_num &lt;- row.names(mydat)
head(mydat)

# plot Vaccmyrt coverage against the CA1 score
gf_point(Vaccmyrt ~ CA1, data=mydat) %&gt;%
  gf_text(Vaccmyrt ~ CA1, label = ~site_num, size = 2, nudge_x = 0.1, nudge_y = 0.1)

# plot Cladstel coverage against the CA1 score
gf_point(Cladstel ~ CA1, data=mydat) %&gt;%
  gf_text(Cladstel ~ CA1, label = ~site_num, size = 2, nudge_x = 0.1, nudge_y = 0.1)</code></pre>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<p>We can see that for Cladstel there are high negative CA1 scores for
sites 9 and 10 which decrease to approximately 0 at site 28. This shows
us that Cladstel contribution to the community is decreasing along the
CA1 axis. In comparison, the relationship between coverage of Vaccmyrt’s
and CA1 scores is the opposite where it goes from high positive CA1
scores at site 28 to approximately zero at sites 9 and 10. If you look
at the figure which plots the species and site scores together, Vaccmyrt
is found close to 28 and Cladsetl is found close to 9 and 10, meaning
that they are important components of those of those sites.</p>
</div>
<div id="section-non-metric-multidimensional-scaling-nmds"
class="section level2">
<h2>Non-metric multidimensional scaling (NMDS)</h2>
<p>Recall that non-metric multidimensional scaling (NMDS) uses a
somewhat different approach to creating the ordination, as it is
concerned with the relative rankings of your sites along each axis,
rather than absolute multi-dimensional distances. It determines these
rankings from pairwise similarity scores between each pair of samples in
turn, and analysing the resultant matrix (which is triangular, rather
like the distances between cities tables you sometimes find in road
atlases).</p>
<p><code>Vegan</code> uses the <code>metaMDS()</code> function to
undertake NMDS based on a distance matrix. It calculates the
<strong>Bray-Curtis</strong> similarity which is robust for most data.
The algorithm has to run multiple times to find the best solution. If
needed, it will automatically standardise your data by either
square-root and/or ‘Wisconsin’ standardisation. You may see this
displayed in the output as the model runs.</p>
<p><strong>Note</strong></p>
<ul>
<li>Use the <code>print()</code> rather than <code>summary()</code>
function to look at the output.</li>
<li>NMDS does not provide you with % variation explained on each axis.
Instead it provides an overall “stress” level, where the lower the
stress, the better.</li>
</ul>
<div class="tutorial-exercise" data-label="metaNMDS" data-completion="1"
data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># Create the NMDS
dune_nmds &lt;- metaMDS(dune)

# Check the output
print(dune_nmds)

# Plot the NMDS sample (site) and attribute (species) scores
plot(dune_nmds)

# plot the site scores
plot(dune_nmds, display = &quot;sites&quot;)

# plot the species scores
plot(dune_nmds, display = &quot;species&quot;)</code></pre>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<p>The stress value is a measure of the goodness of fit. It is important
because it gives an indication of how hard the algorithm is working to
fit these data into 2-dimensional space whilst maximising the
dissimilarities between samples. Stress values greater than &gt;0.3
indicate that the positions in the plot are no better than arbitrary
while anything &lt;0.2 is worth looking at further. A good rule of thumb
for understanding how much you can interpret from your NMDS based on
stress is:</p>
<ul>
<li>stress less than 0.05 indicates an excellent configuration</li>
<li>stress between 0.05 and 0.1 indicates a good configuration</li>
<li>stress between 0.1 and 0.2 requires careful interpretation</li>
<li>stress between 0.2 and 0.3 interpretation becomes difficult</li>
<li>stress greater than 0.3 indicates there is unlikely to be groupings
and the configuration plotted is arbitrary.</li>
</ul>
<p>The NMDS plots with both the <code>sites</code> and
<code>species</code> scores together is difficult to interpret because
we don’t know which species is which or which site is which. We can edit
this plot so that we plot text rather than symbols which might help our
interpretation.</p>
<div class="tutorial-exercise" data-label="metaNMDS2"
data-completion="1" data-diagnostics="1" data-startover="1"
data-lines="0">
<pre class="text"><code># plot empty space first
plot(dune_nmds, type = &quot;n&quot;)

# add text for the sites
text(dune_nmds, display = &quot;sites&quot;, cex = 0.7, col = &quot;blue&quot;)

# add text for the species
text(dune_nmds, display = &quot;species&quot;, cex = 0.7, col = &quot;red&quot;)</code></pre>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
<p>How do we interpret the final plot? Well sites close together will
have similar community composition, in terms of species present and
their abundances. So sites 14, 15 and 20 will have similar community
structure and will be different to 5, 6 and 10.</p>
<p>NMDS is considered to be the most robust unconstrained ordination
method for many biologists. See the paper by Minchin (1987)
<a href="www/Minchin.pdf" target="_blank">here</a></p>
</div>
<div id="section-summary" class="section level2">
<h2>Summary</h2>
<p><strong>Key points</strong></p>
<ul>
<li>Ordination methods provide a convenient way of summarising as much
variation in your raw data into one or two new variables.</li>
<li>Samples close together on these axes will tend to have similar
attributes</li>
<li>It is up to you to try and understand what these axes represent by
interpreting them in relation to environmental data or laboratory
experimental treatments.</li>
</ul>
<p>There is a lot more that you can do with the <code>vegan</code>
package when it comes to analysing community data. On the R website
where the package is hosted, there is more information about the
capabilities of <code>vegan</code> in the <a
href="https://cran.r-project.org/web/packages/vegan/index.html">vignettes
section</a>. It is worthwhile having a look at these if you are
interested to learn more. 
<script type="application/shiny-prerendered" data-context="server-start">
library(learnr)
library(mosaic)
library(vegan)
library(tibble)
dune <- read.csv("www/dune.csv", row.names=1)
data(dune.env)
data(varespec)
dune_pca <- rda(dune)
varespec_ca <- cca(varespec)
dune_nmds <- metaMDS(dune)
knitr::opts_chunk$set(echo = FALSE)
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::register_http_handlers(session, metadata = NULL)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::prepare_tutorial_state(session)
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::i18n_observe_tutorial_language(input, session)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
session$onSessionEnded(function() {
        learnr:::event_trigger(session, "session_stop")
      })
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::question_prerendered_chunk(structure(list(type = "learnr_radio", label = "why_multivariate", 
    question = structure("Why would it be wrong to do lots of linear models, or GLMs, on these\ntypes of data?", html = TRUE, class = c("html", 
    "character")), answers = list(structure(list(id = "lnr_ans_96822c7", 
        option = "It would take too long", value = "It would take too long", 
        label = structure("It would take too long", html = TRUE, class = c("html", 
        "character")), correct = FALSE, message = structure("In R you can automate\nanalyses so the time taken is not a problem.", html = TRUE, class = c("html", 
        "character")), type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_1dd7ad9", 
        option = "There might be more attributes than samples", 
        value = "There might be more attributes than samples", 
        label = structure("There might be more attributes than samples", html = TRUE, class = c("html", 
        "character")), correct = FALSE, message = structure("This\nmight or might not be true, but is not the key issue as it does\nnot affect a linear model or GLM", html = TRUE, class = c("html", 
        "character")), type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_eb6826c", 
        option = "The attributes are not independent of each other, and multiple\n                linear models might give misleading p-values", 
        value = "The attributes are not independent of each other, and multiple\n                linear models might give misleading p-values", 
        label = structure("The attributes are not independent of each other, and multiple\nlinear models might give misleading p-values", html = TRUE, class = c("html", 
        "character")), correct = TRUE, message = structure("Good. The attributes may well be correlated with\neach other, whilst doing lots of individual statistical tests\nis likely to result in something being significant eventually\nsince p=0.05 is merely 1 in 20.", html = TRUE, class = c("html", 
        "character")), type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer"))), button_labels = list(submit = structure("<span data-i18n=\"button.questionsubmit\">Submit Answer<\u002fspan>", html = TRUE, class = c("html", 
    "character")), try_again = structure("<span data-i18n=\"button.questiontryagain\">Try Again<\u002fspan>", html = TRUE, class = c("html", 
    "character"))), messages = list(correct = structure("Correct!", html = TRUE, class = c("html", 
    "character")), try_again = structure("Incorrect", html = TRUE, class = c("html", 
    "character")), incorrect = structure("Incorrect", html = TRUE, class = c("html", 
    "character")), message = NULL, post_message = NULL), ids = list(
        answer = "why_multivariate-answer", question = "why_multivariate"), 
    loading = NULL, random_answer_order = FALSE, allow_retry = TRUE, 
    seed = 59693917.4722029, options = list()), class = c("learnr_radio", 
"tutorial_question")), session = session)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-dune_summary-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-dune_summary-code-editor`)), session)
output$`tutorial-exercise-dune_summary-output` <- renderUI({
  `tutorial-exercise-dune_summary-result`()
})
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "dune_summary", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = NULL, chunks = list(list(label = "dune_summary", 
    code = "", opts = list(label = "\"dune_summary\"", exercise = "TRUE"), 
    engine = "r")), code_check = NULL, error_check = NULL, check = NULL, 
    solution = structure(c("summary(dune)", "head(dune)", "nrow(dune)", 
    "ncol(dune)", "rownames(dune)", "colnames(dune)"), chunk_opts = list(
        label = "dune_summary-solution")), tests = NULL, options = list(
        eval = FALSE, echo = TRUE, results = "markup", tidy = FALSE, 
        tidy.opts = NULL, collapse = FALSE, prompt = FALSE, comment = NA, 
        highlight = FALSE, size = "normalsize", background = "#F7F7F7", 
        strip.white = TRUE, cache = 0, cache.path = "Unconstrained_cache/html/", 
        cache.vars = NULL, cache.lazy = TRUE, dependson = NULL, 
        autodep = FALSE, cache.rebuild = FALSE, fig.keep = "high", 
        fig.show = "asis", fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "dune_summary", exercise = TRUE, 
        code = "", out.width.px = 624, out.height.px = 384, params.src = "dune_summary, exercise=TRUE", 
        fig.num = 0L, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::question_prerendered_chunk(structure(list(type = "learnr_checkbox", label = "simple_pca_interpret", 
    question = structure("Study the PCA plots for sites and species, and tick the true answers", html = TRUE, class = c("html", 
    "character")), answers = list(structure(list(id = "lnr_ans_e6cdff1", 
        option = "Lolium perenne (Lolipere) and Agrostis stolonifera (Agrostol)\n                were rarely found together", 
        value = "Lolium perenne (Lolipere) and Agrostis stolonifera (Agrostol)\n                were rarely found together", 
        label = structure("Lolium perenne (Lolipere) and Agrostis stolonifera (Agrostol)\nwere rarely found together", html = TRUE, class = c("html", 
        "character")), correct = TRUE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_3247051", 
        option = "Quadrat numbers 10, 7 and 5 have similar species composition", 
        value = "Quadrat numbers 10, 7 and 5 have similar species composition", 
        label = structure("Quadrat numbers 10, 7 and 5 have similar species composition", html = TRUE, class = c("html", 
        "character")), correct = TRUE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_eab6ae", 
        option = "Quadrats 18, 9 and 4 have similar species composition", 
        value = "Quadrats 18, 9 and 4 have similar species composition", 
        label = structure("Quadrats 18, 9 and 4 have similar species composition", html = TRUE, class = c("html", 
        "character")), correct = FALSE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_42ba04b", 
        option = "Poa trivialis (Poatriv) is most abundant in quadrats 2, 3, 4 and 9", 
        value = "Poa trivialis (Poatriv) is most abundant in quadrats 2, 3, 4 and 9", 
        label = structure("Poa trivialis (Poatriv) is most abundant in quadrats 2, 3, 4 and 9", html = TRUE, class = c("html", 
        "character")), correct = TRUE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_c49dc8b", 
        option = "Alopecuris geniculatus (Alopgeni) is most abundant in quadrats\n                18, 17 and 19", 
        value = "Alopecuris geniculatus (Alopgeni) is most abundant in quadrats\n                18, 17 and 19", 
        label = structure("Alopecuris geniculatus (Alopgeni) is most abundant in quadrats\n18, 17 and 19", html = TRUE, class = c("html", 
        "character")), correct = FALSE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_7e927f7", 
        option = "Species clustered round both 0 on PC1 and 0 on PC2 at the\n                centre of the plot are fairly ubiquitous across all quadrats", 
        value = "Species clustered round both 0 on PC1 and 0 on PC2 at the\n                centre of the plot are fairly ubiquitous across all quadrats", 
        label = structure("Species clustered round both 0 on PC1 and 0 on PC2 at the\ncentre of the plot are fairly ubiquitous across all quadrats", html = TRUE, class = c("html", 
        "character")), correct = TRUE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer"))), button_labels = list(submit = structure("<span data-i18n=\"button.questionsubmit\">Submit Answer<\u002fspan>", html = TRUE, class = c("html", 
    "character")), try_again = structure("<span data-i18n=\"button.questiontryagain\">Try Again<\u002fspan>", html = TRUE, class = c("html", 
    "character"))), messages = list(correct = structure("Correct!", html = TRUE, class = c("html", 
    "character")), try_again = structure("Incorrect. Be sure to select every correct answer.", html = TRUE, class = c("html", 
    "character")), incorrect = structure("Incorrect", html = TRUE, class = c("html", 
    "character")), message = NULL, post_message = NULL), ids = list(
        answer = "simple_pca_interpret-answer", question = "simple_pca_interpret"), 
    loading = NULL, random_answer_order = FALSE, allow_retry = FALSE, 
    seed = 1617869828.24662, options = list()), class = c("learnr_checkbox", 
"tutorial_question")), session = session)
</script>


<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-sum_pca-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-sum_pca-code-editor`)), session)
output$`tutorial-exercise-sum_pca-output` <- renderUI({
  `tutorial-exercise-sum_pca-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "sum_pca", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = NULL, chunks = list(list(label = "sum_pca", 
    code = "scores(dune_pca, display=\"species\", choices = 1:2)", 
    opts = list(label = "\"sum_pca\"", exercise = "TRUE"), engine = "r")), 
    code_check = NULL, error_check = NULL, check = NULL, solution = NULL, 
    tests = NULL, options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 0, 
        cache.path = "Unconstrained_cache/html/", cache.vars = NULL, 
        cache.lazy = TRUE, dependson = NULL, autodep = FALSE, 
        cache.rebuild = FALSE, fig.keep = "high", fig.show = "asis", 
        fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "sum_pca", exercise = TRUE, 
        code = "scores(dune_pca, display=\"species\", choices = 1:2)", 
        out.width.px = 624, out.height.px = 384, params.src = "sum_pca, exercise = TRUE", 
        fig.num = 0, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>


<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-check_env-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-check_env-code-editor`)), session)
output$`tutorial-exercise-check_env-output` <- renderUI({
  `tutorial-exercise-check_env-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "check_env", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = NULL, chunks = list(list(label = "check_env", 
    code = "", opts = list(label = "\"check_env\"", exercise = "TRUE"), 
    engine = "r")), code_check = NULL, error_check = NULL, check = NULL, 
    solution = structure(c("summary(dune.env)", "head(dune.env)", 
    "nrow(dune.env)", "ncol(dune.env)"), chunk_opts = list(label = "check_env-solution")), 
    tests = NULL, options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 0, 
        cache.path = "Unconstrained_cache/html/", cache.vars = NULL, 
        cache.lazy = TRUE, dependson = NULL, autodep = FALSE, 
        cache.rebuild = FALSE, fig.keep = "high", fig.show = "asis", 
        fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "check_env", exercise = TRUE, 
        code = "", out.width.px = 624, out.height.px = 384, params.src = "check_env, exercise=TRUE", 
        fig.num = 0L, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>


<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-dune_moisture_boxplot-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-dune_moisture_boxplot-code-editor`)), session)
output$`tutorial-exercise-dune_moisture_boxplot-output` <- renderUI({
  `tutorial-exercise-dune_moisture_boxplot-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "dune_moisture_boxplot", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = "dune_pca <- rda(dune)", chunks = list(
    list(label = "dune_moisture_boxplot-setup", code = "dune_pca <- rda(dune)", 
        opts = list(label = "\"dune_moisture_boxplot-setup\""), 
        engine = "r"), list(label = "dune_moisture_boxplot", 
        code = "", opts = list(label = "\"dune_moisture_boxplot\"", 
            exercise = "TRUE"), engine = "r")), code_check = NULL, 
    error_check = NULL, check = NULL, solution = structure(c("# Extract the PC1 site scores", 
    "dune_pc1 <- scores(dune_pca, display=\"sites\", choices = 1)", 
    "", "# Create boxplot", "gf_boxplot(dune_pc1 ~ Moisture, data=dune.env)"
    ), chunk_opts = list(label = "dune_moisture_boxplot-solution")), 
    tests = NULL, options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 0, 
        cache.path = "Unconstrained_cache/html/", cache.vars = NULL, 
        cache.lazy = TRUE, dependson = NULL, autodep = FALSE, 
        cache.rebuild = FALSE, fig.keep = "high", fig.show = "asis", 
        fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "dune_moisture_boxplot", exercise = TRUE, 
        code = "", out.width.px = 624, out.height.px = 384, params.src = "dune_moisture_boxplot, exercise=TRUE", 
        fig.num = 0L, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::question_prerendered_chunk(structure(list(type = "learnr_checkbox", label = "which_spp_like_moisture", 
    question = structure("Which species are particularly associated with wetter soils?", html = TRUE, class = c("html", 
    "character")), answers = list(structure(list(id = "lnr_ans_5cd5625", 
        option = "Lolium perenne (Lolipere)", value = "Lolium perenne (Lolipere)", 
        label = structure("Lolium perenne (Lolipere)", html = TRUE, class = c("html", 
        "character")), correct = FALSE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_fcb6b8", 
        option = "Poa trivialis (Poatriv)", value = "Poa trivialis (Poatriv)", 
        label = structure("Poa trivialis (Poatriv)", html = TRUE, class = c("html", 
        "character")), correct = FALSE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_9ab7787", 
        option = "Poa pratense (Poaprat)", value = "Poa pratense (Poaprat)", 
        label = structure("Poa pratense (Poaprat)", html = TRUE, class = c("html", 
        "character")), correct = FALSE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_d9045aa", 
        option = "Elymus repens (Elymrepe)", value = "Elymus repens (Elymrepe)", 
        label = structure("Elymus repens (Elymrepe)", html = TRUE, class = c("html", 
        "character")), correct = FALSE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_c50e56b", 
        option = "Plantago lanceolata (Planlanc)", value = "Plantago lanceolata (Planlanc)", 
        label = structure("Plantago lanceolata (Planlanc)", html = TRUE, class = c("html", 
        "character")), correct = FALSE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_df126a6", 
        option = "Eleocharis palustris (Eleopalu)", value = "Eleocharis palustris (Eleopalu)", 
        label = structure("Eleocharis palustris (Eleopalu)", html = TRUE, class = c("html", 
        "character")), correct = TRUE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer")), structure(list(id = "lnr_ans_3561231", 
        option = "Agrostis stolonifera (Agrostol)", value = "Agrostis stolonifera (Agrostol)", 
        label = structure("Agrostis stolonifera (Agrostol)", html = TRUE, class = c("html", 
        "character")), correct = TRUE, message = NULL, type = "literal"), class = c("tutorial_question_answer", 
    "tutorial_quiz_answer"))), button_labels = list(submit = structure("<span data-i18n=\"button.questionsubmit\">Submit Answer<\u002fspan>", html = TRUE, class = c("html", 
    "character")), try_again = structure("<span data-i18n=\"button.questiontryagain\">Try Again<\u002fspan>", html = TRUE, class = c("html", 
    "character"))), messages = list(correct = structure("Correct!", html = TRUE, class = c("html", 
    "character")), try_again = structure("Incorrect. Be sure to select every correct answer.", html = TRUE, class = c("html", 
    "character")), incorrect = structure("Incorrect", html = TRUE, class = c("html", 
    "character")), message = NULL, post_message = NULL), ids = list(
        answer = "which_spp_like_moisture-answer", question = "which_spp_like_moisture"), 
    loading = NULL, random_answer_order = TRUE, allow_retry = FALSE, 
    seed = 1874481166.12713, options = list()), class = c("learnr_checkbox", 
"tutorial_question")), session = session)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-check_other_variables-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-check_other_variables-code-editor`)), session)
output$`tutorial-exercise-check_other_variables-output` <- renderUI({
  `tutorial-exercise-check_other_variables-result`()
})
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "check_other_variables", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = "dune_pca <- rda(dune)\ndune_pc1 <- scores(dune_pca, display=\"sites\", choices=1)", 
    chunks = list(list(label = "check_other_variables-setup", 
        code = "dune_pca <- rda(dune)\ndune_pc1 <- scores(dune_pca, display=\"sites\", choices=1)", 
        opts = list(label = "\"check_other_variables-setup\""), 
        engine = "r"), list(label = "check_other_variables", 
        code = "", opts = list(label = "\"check_other_variables\"", 
            exercise = "TRUE"), engine = "r")), code_check = NULL, 
    error_check = NULL, check = NULL, solution = structure(c("gf_point(dune_pc1 ~ A1, data=dune.env)", 
    "gf_boxplot(dune_pc1 ~ Management, data=dune.env)", "gf_boxplot(dune_pc1 ~ Use, data=dune.env)", 
    "gf_boxplot(dune_pc1 ~ Manure, data=dune.env)"), chunk_opts = list(
        label = "check_other_variables-solution")), tests = NULL, 
    options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 0, 
        cache.path = "Unconstrained_cache/html/", cache.vars = NULL, 
        cache.lazy = TRUE, dependson = NULL, autodep = FALSE, 
        cache.rebuild = FALSE, fig.keep = "high", fig.show = "asis", 
        fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "check_other_variables", exercise = TRUE, 
        code = "", out.width.px = 624, out.height.px = 384, params.src = "check_other_variables, exercise=TRUE", 
        fig.num = 0L, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-archpca-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-archpca-code-editor`)), session)
output$`tutorial-exercise-archpca-output` <- renderUI({
  `tutorial-exercise-archpca-result`()
})
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "archpca", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = NULL, chunks = list(list(label = "archpca", 
    code = "# Create the PCA\nvarespec_pca <- rda(varespec)\n\n# Check variation explained; if using RStudio just enter summary(varespec_pca)\nsummary(varespec_pca)[[\"cont\"]][[\"importance\"]]\n\n# Plot PC1 and PC2 site scores only\nplot(varespec_pca, display = \"sites\") \n\n#plot PC1 and PC2 species scores only\nplot(varespec_pca, display = \"species\")", 
    opts = list(label = "\"archpca\"", exercise = "TRUE"), engine = "r")), 
    code_check = NULL, error_check = NULL, check = NULL, solution = NULL, 
    tests = NULL, options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 0, 
        cache.path = "Unconstrained_cache/html/", cache.vars = NULL, 
        cache.lazy = TRUE, dependson = NULL, autodep = FALSE, 
        cache.rebuild = FALSE, fig.keep = "high", fig.show = "asis", 
        fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "archpca", exercise = TRUE, 
        code = c("# Create the PCA", "varespec_pca <- rda(varespec)", 
        "", "# Check variation explained; if using RStudio just enter summary(varespec_pca)", 
        "summary(varespec_pca)[[\"cont\"]][[\"importance\"]]", 
        "", "# Plot PC1 and PC2 site scores only", "plot(varespec_pca, display = \"sites\") ", 
        "", "#plot PC1 and PC2 species scores only", "plot(varespec_pca, display = \"species\")"
        ), out.width.px = 624, out.height.px = 384, params.src = " archpca, exercise=TRUE", 
        fig.num = 0, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-ca1-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-ca1-code-editor`)), session)
output$`tutorial-exercise-ca1-output` <- renderUI({
  `tutorial-exercise-ca1-result`()
})
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "ca1", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = NULL, chunks = list(list(label = "ca1", 
    code = "\n# Create the CA\nvarespec_ca <- cca(varespec)\n\n# Check variation explained; if using RStudio just enter summary(varespec_pca)\nsummary(varespec_ca)\n\n# plot CA1 and CA2 for samples and species\nplot(varespec_ca)\n\n# Plot CA1 and CA2 of the samples\nplot(varespec_ca, display=\"sites\") \n\n# Plot CA1 and CA2 of the species\nplot(varespec_ca, display=\"species\") ", 
    opts = list(label = "\"ca1\"", exercise = "TRUE"), engine = "r")), 
    code_check = NULL, error_check = NULL, check = NULL, solution = NULL, 
    tests = NULL, options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 0, 
        cache.path = "Unconstrained_cache/html/", cache.vars = NULL, 
        cache.lazy = TRUE, dependson = NULL, autodep = FALSE, 
        cache.rebuild = FALSE, fig.keep = "high", fig.show = "asis", 
        fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "ca1", exercise = TRUE, code = c("", 
        "# Create the CA", "varespec_ca <- cca(varespec)", "", 
        "# Check variation explained; if using RStudio just enter summary(varespec_pca)", 
        "summary(varespec_ca)", "", "# plot CA1 and CA2 for samples and species", 
        "plot(varespec_ca)", "", "# Plot CA1 and CA2 of the samples", 
        "plot(varespec_ca, display=\"sites\") ", "", "# Plot CA1 and CA2 of the species", 
        "plot(varespec_ca, display=\"species\") "), out.width.px = 624, 
        out.height.px = 384, params.src = "ca1, exercise=TRUE", 
        fig.num = 0, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-ca3-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-ca3-code-editor`)), session)
output$`tutorial-exercise-ca3-output` <- renderUI({
  `tutorial-exercise-ca3-result`()
})
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "ca3", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = NULL, chunks = list(list(label = "ca3", 
    code = "\n# Extract just the CA1 scores\nsco1 <- scores(varespec_ca, display = \"sites\", ch=1)\n\n# look at the top 6 rows of data\nhead(sco1)\n\n# Select four species of interest and plot against CA1\nvarspec2 <- varespec  %>%  select(Cladstel, Pleuschr, Vaccmyrt, Empenigr)\n\n# Merge CA1 scores and species abundance into one table and turn the row names into a column so we can plot the site numbers on the figure\nmydat <- cbind(sco1, varspec2)\nmydat$site_num <- row.names(mydat)\nhead(mydat)\n\n# plot Vaccmyrt coverage against the CA1 score\ngf_point(Vaccmyrt ~ CA1, data=mydat) %>%\n  gf_text(Vaccmyrt ~ CA1, label = ~site_num, size = 2, nudge_x = 0.1, nudge_y = 0.1)\n\n# plot Cladstel coverage against the CA1 score\ngf_point(Cladstel ~ CA1, data=mydat) %>%\n  gf_text(Cladstel ~ CA1, label = ~site_num, size = 2, nudge_x = 0.1, nudge_y = 0.1)\n", 
    opts = list(label = "\"ca3\"", exercise = "TRUE"), engine = "r")), 
    code_check = NULL, error_check = NULL, check = NULL, solution = NULL, 
    tests = NULL, options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 0, 
        cache.path = "Unconstrained_cache/html/", cache.vars = NULL, 
        cache.lazy = TRUE, dependson = NULL, autodep = FALSE, 
        cache.rebuild = FALSE, fig.keep = "high", fig.show = "asis", 
        fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "ca3", exercise = TRUE, code = c("", 
        "# Extract just the CA1 scores", "sco1 <- scores(varespec_ca, display = \"sites\", ch=1)", 
        "", "# look at the top 6 rows of data", "head(sco1)", 
        "", "# Select four species of interest and plot against CA1", 
        "varspec2 <- varespec  %>%  select(Cladstel, Pleuschr, Vaccmyrt, Empenigr)", 
        "", "# Merge CA1 scores and species abundance into one table and turn the row names into a column so we can plot the site numbers on the figure", 
        "mydat <- cbind(sco1, varspec2)", "mydat$site_num <- row.names(mydat)", 
        "head(mydat)", "", "# plot Vaccmyrt coverage against the CA1 score", 
        "gf_point(Vaccmyrt ~ CA1, data=mydat) %>%", "  gf_text(Vaccmyrt ~ CA1, label = ~site_num, size = 2, nudge_x = 0.1, nudge_y = 0.1)", 
        "", "# plot Cladstel coverage against the CA1 score", 
        "gf_point(Cladstel ~ CA1, data=mydat) %>%", "  gf_text(Cladstel ~ CA1, label = ~site_num, size = 2, nudge_x = 0.1, nudge_y = 0.1)", 
        ""), out.width.px = 624, out.height.px = 384, params.src = "ca3, exercise=TRUE", 
        fig.num = 0, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-metaNMDS-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-metaNMDS-code-editor`)), session)
output$`tutorial-exercise-metaNMDS-output` <- renderUI({
  `tutorial-exercise-metaNMDS-result`()
})
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "metaNMDS", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = NULL, chunks = list(list(label = "metaNMDS", 
    code = "\n# Create the NMDS\ndune_nmds <- metaMDS(dune)\n\n# Check the output\nprint(dune_nmds)\n\n# Plot the NMDS sample (site) and attribute (species) scores\nplot(dune_nmds)\n\n# plot the site scores\nplot(dune_nmds, display = \"sites\")\n\n# plot the species scores\nplot(dune_nmds, display = \"species\")", 
    opts = list(label = "\"metaNMDS\"", exercise = "TRUE"), engine = "r")), 
    code_check = NULL, error_check = NULL, check = NULL, solution = NULL, 
    tests = NULL, options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 0, 
        cache.path = "Unconstrained_cache/html/", cache.vars = NULL, 
        cache.lazy = TRUE, dependson = NULL, autodep = FALSE, 
        cache.rebuild = FALSE, fig.keep = "high", fig.show = "asis", 
        fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "metaNMDS", exercise = TRUE, 
        code = c("", "# Create the NMDS", "dune_nmds <- metaMDS(dune)", 
        "", "# Check the output", "print(dune_nmds)", "", "# Plot the NMDS sample (site) and attribute (species) scores", 
        "plot(dune_nmds)", "", "# plot the site scores", "plot(dune_nmds, display = \"sites\")", 
        "", "# plot the species scores", "plot(dune_nmds, display = \"species\")"
        ), out.width.px = 624, out.height.px = 384, params.src = "metaNMDS, exercise=TRUE", 
        fig.num = 0, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-metaNMDS2-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-metaNMDS2-code-editor`)), session)
output$`tutorial-exercise-metaNMDS2-output` <- renderUI({
  `tutorial-exercise-metaNMDS2-result`()
})
</script>


<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "metaNMDS2", global_setup = structure(c("library(learnr)", 
"library(mosaic)", "library(vegan)", "library(tibble)", "dune <- read.csv(\"www/dune.csv\", row.names=1)", 
"data(dune.env)", "data(varespec)", "dune_pca <- rda(dune)", 
"varespec_ca <- cca(varespec)", "dune_nmds <- metaMDS(dune)", 
"knitr::opts_chunk$set(echo = FALSE)"), chunk_opts = list(label = "setup", 
    include = FALSE)), setup = NULL, chunks = list(list(label = "metaNMDS2", 
    code = "# plot empty space first\nplot(dune_nmds, type = \"n\")\n\n# add text for the sites\ntext(dune_nmds, display = \"sites\", cex = 0.7, col = \"blue\")\n\n# add text for the species\ntext(dune_nmds, display = \"species\", cex = 0.7, col = \"red\")\n", 
    opts = list(label = "\"metaNMDS2\"", exercise = "TRUE"), 
    engine = "r")), code_check = NULL, error_check = NULL, check = NULL, 
    solution = NULL, tests = NULL, options = list(eval = FALSE, 
        echo = TRUE, results = "markup", tidy = FALSE, tidy.opts = NULL, 
        collapse = FALSE, prompt = FALSE, comment = NA, highlight = FALSE, 
        size = "normalsize", background = "#F7F7F7", strip.white = TRUE, 
        cache = 0, cache.path = "Unconstrained_cache/html/", 
        cache.vars = NULL, cache.lazy = TRUE, dependson = NULL, 
        autodep = FALSE, cache.rebuild = FALSE, fig.keep = "high", 
        fig.show = "asis", fig.align = "default", fig.path = "Unconstrained_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 6.5, fig.height = 4, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 624, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        max.print = 1000, label = "metaNMDS2", exercise = TRUE, 
        code = c("# plot empty space first", "plot(dune_nmds, type = \"n\")", 
        "", "# add text for the sites", "text(dune_nmds, display = \"sites\", cex = 0.7, col = \"blue\")", 
        "", "# add text for the species", "text(dune_nmds, display = \"species\", cex = 0.7, col = \"red\")", 
        ""), out.width.px = 624, out.height.px = 384, params.src = "metaNMDS2, exercise = TRUE", 
        fig.num = 0, exercise.df_print = "paged", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
</p>
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="dependencies">
{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["header-attrs"]},{"type":"character","attributes":{},"value":["2.22"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pandoc"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["header-attrs.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.22"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["3.6.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/3.6.0"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery-3.6.0.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquerylib"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.1.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootstrap"]},{"type":"character","attributes":{},"value":["3.3.5"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/bootstrap"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["viewport"]}},"value":[{"type":"character","attributes":{},"value":["width=device-width, initial-scale=1"]}]},{"type":"character","attributes":{},"value":["js/bootstrap.min.js","shim/html5shiv.min.js","shim/respond.min.js"]},{"type":"character","attributes":{},"value":["css/cerulean.min.css"]},{"type":"character","attributes":{},"value":["<style>h1 {font-size: 34px;}\n       h1.title {font-size: 38px;}\n       h2 {font-size: 30px;}\n       h3 {font-size: 24px;}\n       h4 {font-size: 18px;}\n       h5 {font-size: 16px;}\n       h6 {font-size: 12px;}\n       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}\n       pre:not([class]) { background-color: white }<\/style>"]},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.22"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["pagedtable"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pagedtable-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["js/pagedtable.js"]},{"type":"character","attributes":{},"value":["css/pagedtable.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.22"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["textmate.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.22"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.11.4"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["i18n"]},{"type":"character","attributes":{},"value":["21.6.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/i18n"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["i18next.min.js","tutorial-i18n-init.js"]},{"type":"NULL"},{"type":"character","attributes":{},"value":["<script id=\"i18n-cstm-trns\" type=\"application/json\">{\"language\":\"en\",\"resources\":{\"en\":{\"translation\":{\"button\":{\"runcode\":\"Run Code\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"Hint\",\"hint_plural\":\"Hints\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"Next Hint\",\"hintprev\":\"Previous Hint\",\"solution\":\"Solution\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"Copy to Clipboard\",\"startover\":\"Start Over\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"Continue\",\"submitanswer\":\"Submit Answer\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"Previous Topic\",\"nexttopic\":\"Next Topic\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"Try Again\"},\"text\":{\"startover\":\"Start Over\",\"areyousure\":\"Are you sure you want to start over? (all exercise progress will be reset)\",\"youmustcomplete\":\"You must complete the\",\"exercise\":\"exercise\",\"exercise_plural\":\"exercises\",\"inthissection\":\"in this section before continuing.\",\"code\":\"Code\",\"enginecap\":\"{{engine}} $t(text.code)\",\"quiz\":\"Quiz\",\"blank\":\"blank\",\"blank_plural\":\"blanks\",\"exercisecontainsblank\":\"This exercise contains {{count}} $t(text.blank).\",\"pleasereplaceblank\":\"Please replace {{blank}} with valid code.\",\"unparsable\":\"It looks like this might not be valid R code. R cannot determine how to turn your text into a complete command. You may have forgotten to fill in a blank, to remove an underscore, to include a comma between arguments, or to close an opening <code>&quot;<\\/code>, <code>'<\\/code>, <code>(<\\/code> or <code>{<\\/code> with a matching <code>&quot;<\\/code>, <code>'<\\/code>, <code>)<\\/code> or <code>}<\\/code>.\\n\",\"unparsablequotes\":\"<p>It looks like your R code contains specially formatted quotation marks or &quot;curly&quot; quotes (<code>{{character}}<\\/code>) around character strings, making your code invalid. R requires character values to be contained in straight quotation marks (<code>&quot;<\\/code> or <code>'<\\/code>).<\\/p> {{code}} <p>Don't worry, this is a common source of errors when you copy code from another app that applies its own formatting to text. You can try replacing the code on that line with the following. There may be other places that need to be fixed, too.<\\/p> {{suggestion}}\\n\",\"unparsableunicode\":\"<p>It looks like your R code contains an unexpected special character (<code>{{character}}<\\/code>) that makes your code invalid.<\\/p> {{code}} <p>Sometimes your code may contain a special character that looks like a regular character, especially if you copy and paste the code from another app. Try deleting the special character from your code and retyping it manually.<\\/p>\\n\",\"unparsableunicodesuggestion\":\"<p>It looks like your R code contains an unexpected special character (<code>{{character}}<\\/code>) that makes your code invalid.<\\/p> {{code}} <p>Sometimes your code may contain a special character that looks like a regular character, especially if you copy and paste the code from another app. You can try replacing the code on that line with the following. There may be other places that need to be fixed, too.<\\/p> {{suggestion}}\\n\",\"and\":\"and\",\"or\":\"or\",\"listcomma\":\", \",\"oxfordcomma\":\",\"}}},\"fr\":{\"translation\":{\"button\":{\"runcode\":\"Lancer le Code\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"Indication\",\"hint_plural\":\"Indications\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"Indication Suivante\",\"hintprev\":\"Indication Précédente\",\"solution\":\"Solution\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"Copier dans le Presse-papier\",\"startover\":\"Recommencer\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"Continuer\",\"submitanswer\":\"Soumettre\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"Chapitre Précédent\",\"nexttopic\":\"Chapitre Suivant\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"Réessayer\"},\"text\":{\"startover\":\"Recommencer\",\"areyousure\":\"Êtes-vous certains de vouloir recommencer? (La progression sera remise à zéro)\",\"youmustcomplete\":\"Vous devez d'abord compléter\",\"exercise\":\"l'exercice\",\"exercise_plural\":\"des exercices\",\"inthissection\":\"de cette section avec de continuer.\",\"code\":\"Code\",\"enginecap\":\"$t(text.code) {{engine}}\",\"quiz\":\"Quiz\",\"and\":\"et\",\"or\":\"ou\",\"oxfordcomma\":\"\"}}},\"es\":{\"translation\":{\"button\":{\"runcode\":\"Ejecutar código\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"Pista\",\"hint_plural\":\"Pistas\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"Siguiente pista\",\"hintprev\":\"Pista anterior\",\"solution\":\"Solución\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"Copiar al portapapeles\",\"startover\":\"Reiniciar\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"Continuar\",\"submitanswer\":\"Enviar respuesta\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"Tema anterior\",\"nexttopic\":\"Tema siguiente\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"Volver a intentar\"},\"text\":{\"startover\":\"Reiniciar\",\"areyousure\":\"¿De verdad quieres empezar de nuevo? (todo el progreso del ejercicio se perderá)\",\"youmustcomplete\":\"Debes completar\",\"exercise\":\"el ejercicio\",\"exercise_plural\":\"los ejercicios\",\"inthissection\":\"en esta sección antes de continuar.\",\"code\":\"Código\",\"enginecap\":\"$t(text.code) {{engine}}\",\"quiz\":\"Cuestionario\",\"and\":\"y\",\"or\":\"o\",\"oxfordcomma\":\"\"}}},\"pt\":{\"translation\":{\"button\":{\"runcode\":\"Executar código\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"Dica\",\"hint_plural\":\"Dicas\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"Próxima dica\",\"hintprev\":\"Dica anterior\",\"solution\":\"Solução\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"Copiar para a área de transferência\",\"startover\":\"Reiniciar\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"Continuar\",\"submitanswer\":\"Enviar resposta\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"Tópico anterior\",\"nexttopic\":\"Próximo tópico\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"Tentar novamente\"},\"text\":{\"startover\":\"Reiniciar\",\"areyousure\":\"Tem certeza que deseja começar novamente? (todo o progresso feito será perdido)\",\"youmustcomplete\":\"Você deve completar\",\"exercise\":\"o exercício\",\"exercise_plural\":\"os exercícios\",\"inthissection\":\"nesta seção antes de continuar.\",\"code\":\"Código\",\"enginecap\":\"$t(text.code) {{engine}}\",\"quiz\":\"Quiz\",\"and\":\"e\",\"or\":\"ou\",\"oxfordcomma\":\"\"}}},\"tr\":{\"translation\":{\"button\":{\"runcode\":\"Çalıştırma Kodu\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"Ipucu\",\"hint_plural\":\"İpuçları\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"Sonraki İpucu\",\"hintprev\":\"Önceki İpucu\",\"solution\":\"Çözüm\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"Pano'ya Kopyala\",\"startover\":\"Baştan Başlamak\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"Devam et\",\"submitanswer\":\"Cevabı onayla\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"Önceki Konu\",\"nexttopic\":\"Sonraki Konu\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"Tekrar Deneyin\"},\"text\":{\"startover\":\"Baştan Başlamak\",\"areyousure\":\"Baştan başlamak istediğinizden emin misiniz? (tüm egzersiz ilerlemesi kaybolacak)\",\"youmustcomplete\":\"Tamamlamalısın\",\"exercise\":\"egzersiz\",\"exercise_plural\":\"egzersizler\",\"inthissection\":\"devam etmeden önce bu bölümde\",\"code\":\"Kod\",\"enginecap\":\"$t(text.code) {{engine}}\",\"quiz\":\"Sınav\",\"oxfordcomma\":\"\"}}},\"emo\":{\"translation\":{\"button\":{\"runcode\":\"🏃\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"💡\",\"hint_plural\":\"$t(button.hint)\",\"hinttitle\":\"$t(button.hint)\",\"solution\":\"🎯\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"📋\",\"startover\":\"⏮\",\"startovertitle\":\"Start Over\",\"continue\":\"✅\",\"submitanswer\":\"🆗\",\"submitanswertitle\":\"Submit Answer\",\"previoustopic\":\"⬅\",\"nexttopic\":\"➡\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"🔁\"},\"text\":{\"startover\":\"⏮\",\"areyousure\":\"🤔\",\"youmustcomplete\":\"⚠️ 👉 🧑‍💻\",\"exercise\":\"\",\"exercise_plural\":\"\",\"inthissection\":\"\",\"code\":\"💻\",\"enginecap\":\"$t(text.code) {{engine}}\",\"oxfordcomma\":\"\"}}},\"eu\":{\"translation\":{\"button\":{\"runcode\":\"Kodea egikaritu\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"Laguntza\",\"hint_plural\":\"Laguntza\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"Aurreko laguntza\",\"hintprev\":\"Hurrengo laguntza\",\"solution\":\"Ebazpena\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"Arbelean kopiatu\",\"startover\":\"Berrabiarazi\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"Jarraitu\",\"submitanswer\":\"Erantzuna bidali\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"Aurreko atala\",\"nexttopic\":\"Hurrengo atala\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"Berriro saiatu\"},\"text\":{\"startover\":\"Berrabiarazi\",\"areyousure\":\"Berriro hasi nahi duzu? (egindako lana galdu egingo da)\",\"youmustcomplete\":\"Aurrera egin baino lehen atal honetako\",\"exercise\":\"ariketa egin behar duzu.\",\"exercise_plural\":\"ariketak egin behar dituzu.\",\"inthissection\":\"\",\"code\":\"Kodea\",\"enginecap\":\"$t(text.code) {{engine}}\",\"quiz\":\"Galdetegia\",\"oxfordcomma\":\"\"}}},\"de\":{\"translation\":{\"button\":{\"runcode\":\"Code ausführen\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"Tipp\",\"hint_plural\":\"Tipps\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"Nächster Tipp\",\"hintprev\":\"Vorheriger Tipp\",\"solution\":\"Lösung\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"In die Zwischenablage kopieren\",\"startover\":\"Neustart\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"Weiter\",\"submitanswer\":\"Antwort einreichen\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"Vorheriges Kapitel\",\"nexttopic\":\"Nächstes Kapitel\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"Nochmal versuchen\"},\"text\":{\"startover\":\"Neustart\",\"areyousure\":\"Bist du sicher, dass du neustarten willst? (der gesamte Lernfortschritt wird gelöscht)\",\"youmustcomplete\":\"Vervollstädinge\",\"exercise\":\"die Übung\",\"exercise_plural\":\"die Übungen\",\"inthissection\":\"in diesem Kapitel, bevor du fortfährst.\",\"code\":\"Code\",\"enginecap\":\"$t(text.code) {{engine}}\",\"quiz\":\"Quiz\",\"blank\":\"Lücke\",\"blank_plural\":\"Lücken\",\"pleasereplaceblank\":\"Bitte ersetze {{blank}} mit gültigem Code.\",\"unparsable\":\"Dies scheint kein gültiger R Code zu sein. R kann deinen Text nicht in einen gültigen Befehl übersetzen. Du hast vielleicht vergessen, die Lücke zu füllen, einen Unterstrich zu entfernen, ein Komma zwischen Argumente zu setzen oder ein eröffnendes <code>&quot;<\\/code>, <code>'<\\/code>, <code>(<\\/code> oder <code>{<\\/code> mit einem zugehörigen <code>&quot;<\\/code>, <code>'<\\/code>, <code>)<\\/code> oder <code>}<\\/code> zu schließen.\\n\",\"and\":\"und\",\"or\":\"oder\",\"listcomma\":\", \",\"oxfordcomma\":\",\"}}},\"ko\":{\"translation\":{\"button\":{\"runcode\":\"코드 실행\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"힌트\",\"hint_plural\":\"힌트들\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"다음 힌트\",\"hintprev\":\"이전 힌트\",\"solution\":\"솔루션\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"클립보드에 복사\",\"startover\":\"재학습\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"다음 학습으로\",\"submitanswer\":\"정답 제출\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"이전 토픽\",\"nexttopic\":\"다음 토픽\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"재시도\"},\"text\":{\"startover\":\"재학습\",\"areyousure\":\"다시 시작 하시겠습니까? (모든 예제의 진행 정보가 재설정됩니다)\",\"youmustcomplete\":\"당신은 완료해야 합니다\",\"exercise\":\"연습문제\",\"exercise_plural\":\"연습문제들\",\"inthissection\":\"이 섹션을 실행하기 전에\",\"code\":\"코드\",\"enginecap\":\"$t(text.code) {{engine}}\",\"quiz\":\"퀴즈\",\"blank\":\"공백\",\"blank_plural\":\"공백들\",\"exercisecontainsblank\":\"이 연습문제에는 {{count}}개의 $t(text.blank)이 포함되어 있습니다.\",\"pleasereplaceblank\":\"{{blank}}를 유효한 코드로 바꾸십시오.\",\"unparsable\":\"이것은 유효한 R 코드가 아닐 수 있습니다. R은 텍스트를 완전한 명령으로 변환하는 방법을 결정할 수 없습니다. 당신은 공백이나 밑줄을 대체하여 채우기, 인수를 컴마로 구분하기, 또는 <code>&quot;<\\/code>, <code>'<\\/code>, <code>(<\\/code> , <code>{<\\/code>로 시작하는 구문을 닫는 <code>&quot;<\\/code>, <code>'<\\/code>, <code>)<\\/code>, <code>}<\\/code>을 잊었을 수도 있습니다.\\n\",\"and\":\"그리고\",\"or\":\"혹은\",\"listcomma\":\", \",\"oxfordcomma\":\"\"}}},\"zh\":{\"translation\":{\"button\":{\"runcode\":\"运行代码\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"提示\",\"hint_plural\":\"提示\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"下一个提示\",\"hintprev\":\"上一个提示\",\"solution\":\"答案\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"复制到剪切板\",\"startover\":\"重新开始\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"继续\",\"submitanswer\":\"提交答案\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"上一专题\",\"nexttopic\":\"下一专题\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"再试一次\"},\"text\":{\"startover\":\"重置\",\"areyousure\":\"你确定要重新开始吗? (所有当前进度将被重置)\",\"youmustcomplete\":\"你必须完成\",\"exercise\":\"练习\",\"exercise_plural\":\"练习\",\"inthissection\":\"在进行本节之前\",\"code\":\"代码\",\"enginecap\":\"$t(text.code) {{engine}}\",\"quiz\":\"测试\",\"blank\":\"空\",\"blank_plural\":\"空\",\"exercisecontainsblank\":\"本练习包含{{count}}个$t(text.blank)\",\"pleasereplaceblank\":\"请在{{blank}}内填写恰当的代码\",\"unparsable\":\"这似乎不是有效的R代码。 R不知道如何将您的文本转换为完整的命令。 您是否忘了填空，忘了删除下划线，忘了在参数之间包含逗号，或者是忘了用<code>&quot;<\\/code>, <code>'<\\/code>, <code>)<\\/code>,<code>}<\\/code>来封闭<code>&quot;<\\/code>, <code>'<\\/code>, <code>(<\\/code>。 or <code>{<\\/code>。\\n\",\"unparsablequotes\":\"<p>您的R代码中似乎含有特殊格式的引号，或者弯引号(<code>{{character}}<\\/code>) 在字符串前后，在R中字符串应该被直引号(<code>&quot;<\\/code> 或者 <code>'<\\/code>)包裹。<\\/p> {{code}} <p>别担心，该错误经常在复制粘贴包含格式的代码时遇到， 您可以尝试将该行中的代码替换为以下代码，也许还有其他地方需要修改。<\\/p> {{suggestion}}\\n\",\"unparsableunicode\":\"<p>您的代码中似乎包含有异常字符(<code>{{character}}<\\/code>),导致代码无效。<\\/p> {{code}} <p>有时候你的代码可能含有看似正常字符的特殊字符，特别是当你复制粘贴其他来源代码的时候。 请试着删除这些特殊字符,重新输入<\\/p>\\n\",\"unparsableunicodesuggestion\":\"<p>您的代码中似乎包含有异常字符(<code>{{character}}<\\/code>),导致代码无效。<\\/p> {{code}} <p>有时候你的代码可能含有看似正常字符的特殊字符，特别是当你复制粘贴其他来源代码的时候。 请试着删除这些特殊字符,重新输入<\\/p>\\n\",\"and\":\"且\",\"or\":\"或\",\"listcomma\":\",\",\"oxfordcomma\":\",\"}}},\"pl\":{\"translation\":{\"button\":{\"runcode\":\"Uruchom kod\",\"runcodetitle\":\"$t(button.runcode) ({{kbd}})\",\"hint\":\"Podpowiedź\",\"hint_plural\":\"Podpowiedzi\",\"hinttitle\":\"$t(button.hint)\",\"hintnext\":\"Następna podpowiedź\",\"hintprev\":\"Poprzednia podpowiedź\",\"solution\":\"Rozwiązanie\",\"solutiontitle\":\"$t(button.solution)\",\"copyclipboard\":\"Kopiuj do schowka\",\"startover\":\"Zacznij od początku\",\"startovertitle\":\"$t(button.startover)\",\"continue\":\"Kontynuuj\",\"submitanswer\":\"Wyślij\",\"submitanswertitle\":\"$t(button.submitanswer)\",\"previoustopic\":\"Poprzednia sekcja\",\"nexttopic\":\"Następna sekcja\",\"questionsubmit\":\"$t(button.submitanswer)\",\"questiontryagain\":\"Spróbuj ponownie\"},\"text\":{\"startover\":\"Zacznij od początku\",\"areyousure\":\"Czy na pewno chcesz zacząć od początku? (cały postęp w zadaniu zostanie utracony)\",\"youmustcomplete\":\"Musisz ukończyć\",\"exercise\":\"ćwiczenie\",\"exercise_plural\":\"ćwiczenia\",\"inthissection\":\"w tej sekcji przed kontynuowaniem\",\"code\":\"Kod\",\"enginecap\":\"$t(text.code) {{engine}}\",\"quiz\":\"Quiz\",\"blank\":\"luka\",\"blank_plural\":\"luk(i)\",\"exercisecontainsblank\":\"To ćwiczenie zawiera {{count}} $t(text.blank).\",\"pleasereplaceblank\":\"Proszę uzupełnić {{blank}} prawidłowym kodem.\",\"unparsable\":\"Wygląda na to, że może to nie być prawidłowy kod R. R nie jest w stanie przetworzyć Twojego tekstu na polecenie. Mogłeś(-aś) zapomnieć wypełnić luki, usunąć podkreślnik, umieścić przecinka między argumentami, lub zamknąć znak <code>&quot;<\\/code>, <code>'<\\/code>, <code>(<\\/code> lub <code>{<\\/code> odpowiadającym <code>&quot;<\\/code>, <code>'<\\/code>, <code>)<\\/code> lub <code>}<\\/code>.\\n\",\"unparsablequotes\":\"<p>Wygląda na to, że Twój kod zawiera szczególnie sformatowane cudzysłowy lub cudzysłowy typograficzne (<code>{{character}}<\\/code>) przy ciągach znaków, co sprawia, że kod jest niepoprawny. R wymaga cudzysłowów prostych (<code>&quot;<\\/code> albo <code>'<\\/code>).<\\/p> {{code}} <p>Nie martw się, to powszechne źródło błędów, gdy kopiuje się kod z innego programu, który sam formatuje teskt. Możesz spróbować zastąpić swój kod następującym kodem. Mogą być też inne miejsca, które wymagają poprawienia.<\\/p> {{suggestion}}\\n\",\"unparsableunicode\":\"<p>Wygląda na to, że Twój kod zawiera niespodziewany znak specjalny (<code>{{character}}<\\/code>), co sprawia, że kod jest niepoprawny.<\\/p> {{code}} <p>Czasami Twój kod może zawierać znak specjalny, który wygląda jak zwykły znak, zwłaszcza jeśli kopiujesz kod z innego programu. Spróbuj usunąć znak specjalny i wpisać do ponownie ręcznie.<\\/p>\\n\",\"unparsableunicodesuggestion\":\"<p>Wygląda na to, że Twój kod zawiera niespodziewany znak specjalny (<code>{{character}}<\\/code>), co sprawia, że kod jest niepoprawny.<\\/p> {{code}} <p>Czasami Twój kod może zawierać znak specjalny, który wygląda jak zwykły znak, zwłaszcza jeśli kopiujesz kod z innego programu. Możesz spróbować zastąpić swój kod następującym kodem. Mogą być też inne miejsca, które wymagają poprawienia.<\\/p> {{suggestion}}\\n\",\"and\":\"i\",\"or\":\"lub\",\"listcomma\":\", \",\"oxfordcomma\":\"\"}}}}}<\/script>"]},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-format"]},{"type":"character","attributes":{},"value":["0.11.4"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmarkdown/templates/tutorial/resources"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-format.js"]},{"type":"character","attributes":{},"value":["tutorial-format.css","rstudio-theme.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["3.6.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/3.6.0"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery-3.6.0.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquerylib"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.1.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["navigation"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/navigation-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tabsets.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.22"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["default.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.22"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["3.6.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/3.6.0"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery-3.6.0.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquerylib"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.1.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["font-awesome"]},{"type":"character","attributes":{},"value":["6.4.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["fontawesome"]}]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["css/all.min.css","css/v4-shims.min.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["fontawesome"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.5.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootbox"]},{"type":"character","attributes":{},"value":["5.5.2"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/bootbox"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["bootbox.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["idb-keyvalue"]},{"type":"character","attributes":{},"value":["3.2.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/idb-keyval"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["idb-keyval-iife-compat.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.11.4"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.4"]}]}]}
</script>
<!--/html_preserve-->
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="execution_dependencies">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages","version"]},"class":{"type":"character","attributes":{},"value":["data.frame"]},"row.names":{"type":"integer","attributes":{},"value":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90]}},"value":[{"type":"character","attributes":{},"value":["backports","base","bslib","cachem","checkmate","cli","cluster","colorspace","commonmark","compiler","datasets","digest","dplyr","ellipsis","evaluate","fansi","farver","fastmap","fontawesome","forcats","generics","ggforce","ggformula","ggplot2","ggridges","ggstance","glue","graphics","grDevices","grid","gtable","haven","highr","hms","htmltools","htmlwidgets","httpuv","jquerylib","jsonlite","knitr","labelled","later","lattice","learnr","lifecycle","magrittr","markdown","MASS","Matrix","methods","mgcv","mime","mosaic","mosaicCore","mosaicData","munsell","nlme","parallel","permute","pillar","pkgconfig","polyclip","promises","purrr","R6","Rcpp","rlang","rmarkdown","rprojroot","rstudioapi","sass","scales","shiny","splines","stats","stringi","stringr","tibble","tidyr","tidyselect","tools","tweenr","utf8","utils","vctrs","vegan","withr","xfun","xtable","yaml"]},{"type":"character","attributes":{},"value":["1.4.1","4.3.0","0.5.0","1.0.8","2.2.0","3.6.1","2.1.4","2.1-0","1.9.0","4.3.0","4.3.0","0.6.31","1.1.2","0.3.2","0.21","1.0.4","2.1.1","1.1.1","0.5.1","1.0.0","0.1.3","0.4.1","0.10.4","3.4.2","0.5.4","0.3.6","1.6.2","4.3.0","4.3.0","4.3.0","0.3.3","2.5.2","0.10","1.1.3","0.5.5","1.6.2","1.6.11","0.1.4","1.8.5","1.43","2.12.0","1.3.1","0.21-8","0.11.4","1.0.3","2.0.3","1.7","7.3-58.4","1.5-4","4.3.0","1.8-42","0.12","1.8.4.2","0.9.2.1","0.20.3","0.5.0","3.1-162","4.3.0","0.9-7","1.9.0","2.0.3","1.10-4","1.2.0.1","1.0.1","2.5.1","1.0.10","1.1.1","2.22","2.0.3","0.14","0.4.6","1.2.1","1.7.4","4.3.0","4.3.0","1.7.12","1.5.0","3.2.1","1.3.0","1.2.0","4.3.0","2.0.2","1.2.3","4.3.0","0.6.3","2.6-4","2.5.0","0.39","1.8-4","2.3.7"]}]}]}
</script>
<!--/html_preserve-->
</div>

</article> <!-- topics -->

<div class="topicsContainer">
<div class="topicsPositioner">
<div class="band">
<div class="bandContent topicsListContainer">

<!-- begin doc-metadata -->
<div id="doc-metadata">
<h1 class="title toc-ignore" style="display:none;">Multiple response
variables</h1>
</div>
<!-- end doc-metadata -->

</div> <!-- bandContent.topicsListContainer -->
</div> <!-- band -->
</div> <!-- topicsPositioner -->
</div> <!-- topicsContainer -->


</main> <!-- bandContent page -->
</div> <!-- pageContent band -->



<!-- Build Tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("section-TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<script>
// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>


</body>

</html>
